all.yml

```
@@ -1,24 +1,31 @@
 ---
-- name: Generate CA for Self Signed Certs
-  hosts: zookeeper[0]
+- name: Host Prerequisites
+  hosts: zookeeper:kafka_broker:schema_registry:kafka_connect:ksql:control_center:kafka_rest
   gather_facts: false
-  tags: certificate_authority
+  environment: "{{ proxy_env }}"
   tasks:
-  - import_role:
-      name: confluent.variables
-
   - name: Create Certificate Authority and Copy to Ansible Host
     include_tasks: tasks/certificate_authority.yml
+    tags: certificate_authority
+    run_once: true
     when: (self_signed|bool and regenerate_ca|bool and kafka_broker_listeners | ssl_required(ssl_enabled)) or create_mds_certs|bool
 
-- name: Host Prerequisites
-  hosts: zookeeper:kafka_broker:schema_registry:kafka_connect:ksql:control_center:kafka_rest
-  gather_facts: false
-  tags: common
-  environment: "{{ proxy_env }}"
-  tasks:
   - import_role:
       name: confluent.common
+    tags: common
+
+  - name: Set Up Secrets Protection
+    include_tasks: tasks/masterkey.yml
+    run_once: true
+    when: secrets_protection_enabled|bool
+
+  - name: Save Master Key to Host
+    set_fact: secrets_protection_masterkey={{hostvars[groups['zookeeper'][0]].secrets_protection_masterkey}}
+    when: (secrets_protection_enabled|bool) and (not secrets_protection_masterkey)
+
+  - name: Save Security File to Host
+    set_fact: secrets_protection_security_file={{hostvars[groups['zookeeper'][0]].secrets_protection_security_file}}
+    when: (secrets_protection_enabled|bool)
```

```suggestion
  - name: Set Up Secrets Protection
    include_tasks: tasks/masterkey.yml
    when: secrets_protection_enabled|bool
```

You should be able to move the run_once logic and the set fact tasks into the masterkey.yml tasks file